storage/innobase/vers/vers0vtq.cc

```diff
@@ -0,0 +1,472 @@
+	heap = mem_heap_create(0);
+
+	ut_ad(index);
+	ut_ad(dict_index_is_clust(index));
+
+	mach_write_to_8(
+		reinterpret_cast<byte*>(&trx_id_net),
+		in_trx_id);
+
+	tuple = dtuple_create(heap, 1);
+	dfield = dtuple_get_nth_field(tuple, DICT_FLD__SYS_VTQ__TRX_ID);
+	dfield_set_data(dfield, &trx_id_net, 8);
+	dict_index_copy_types(tuple, index, 1);
+
+	mtr_start_trx(&mtr, trx);
+	btr_pcur_open_on_user_rec(index, tuple, PAGE_CUR_GE,
+			BTR_SEARCH_LEAF, &pcur, &mtr);
+
+	if (!btr_pcur_is_on_user_rec(&pcur))
+		goto not_found;
+
+	rec = btr_pcur_get_rec(&pcur);
+	{
+		const char *err = trx->vtq_query.cache_result(heap, rec);
+		if (err) {
+			fprintf(stderr, "InnoDB: vtq_query_trx_id: get VTQ field failed: %s\n", err);
+			ut_ad(false && "get VTQ field failed");
+			goto not_found;
+		}
+	}
+
+	if (cached.trx_id != in_trx_id)
+		goto not_found;
+
+	vtq_result(thd, cached, out, field);
+	found = true;
+
+not_found:
+	btr_pcur_close(&pcur);
+	mtr_commit(&mtr);
+	mem_heap_free(heap);
+
+	DBUG_RETURN(found);
+}
+
+static
+inline
+void rec_get_timeval(const rec_t* rec, ulint nfield, timeval& out)
+{
+	ulint		len;
+	const byte*	field;
+	field = rec_get_nth_field_old(
+		rec, nfield, &len);
+
+	ut_ad(len == sizeof(ullong));
+
+	out.tv_sec = mach_read_from_4(field);
+	out.tv_usec = mach_read_from_4(field + 4);
+}
+
+inline
+const char *
+vtq_query_t::cache_result(
+	mem_heap_t* heap,
+	const rec_t* rec,
+	const timeval& _ts_query,
+	bool _backwards)
+{
+	prev_query = _ts_query;
+	backwards = _backwards;
+	return dict_process_sys_vtq(heap, rec, result);
+}
+
+static
+inline
+bool
+operator== (const timeval &a, const timeval &b)
+{
+	return a.tv_sec == b.tv_sec && a.tv_usec == b.tv_usec;
+}
+
+static
+inline
+bool
+operator!= (const timeval &a, const timeval &b)
+{
+	return !(a == b);
+}
+
+static
+inline
+bool
+operator> (const timeval &a, const timeval &b)
+{
+	return a.tv_sec > b.tv_sec || (a.tv_sec == b.tv_sec && a.tv_usec > b.tv_usec);
+}
+
+static
+inline
+bool
+operator< (const timeval &a, const timeval &b)
+{
+	return b > a;
+}
+
+static
+trx_id_t
+read_trx_id(const rec_t *rec) {
+	ulint len = 0;
+	const rec_t *field = rec_get_nth_field_old(rec, 1, &len);
+	DBUG_ASSERT(len == 8);
+	return mach_read_from_8(field);
+}
+
+/** Find a row with given commit_ts but MAX()/MIN() trx_id
+@param[in]	mtr		mini-transaction handler
+@param[in, out]	pcur		btree cursor which may be changed by this function
+@param[in]	backwards	search direction
+@param[in]	commit_ts	target timestamp for records
+@param[in]	rec		record buffer for pcur
+*/
+static
+void
+find_best_match(
+	mtr_t &mtr,
+	btr_pcur_t &pcur,
+	bool backwards,
+	timeval commit_ts,
+	const rec_t *rec)
+{
+	btr_pcur_t best;
+	btr_pcur_init(&best);
+	btr_pcur_copy_stored_position(&best, &pcur);
+	trx_id_t best_trx_id = read_trx_id(rec);
+
+	while (true) {
+		if (backwards ? !btr_pcur_move_to_prev_user_rec(&pcur, &mtr)
+			      : !btr_pcur_move_to_next_user_rec(&pcur, &mtr))
+			break;
```

Comment:
I would suggest moving the `backwards` predicate outside of the loop, and implementing separate loops for each case.